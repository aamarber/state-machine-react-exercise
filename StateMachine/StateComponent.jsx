import React from 'react'

export default function StateComponent({stateSize, state, statePosition, isLast}) {
  
    const topMargin = 10;

    const calcCircleMargin = (circleIndex) => {
        return 150 * circleIndex;
    }

    const calcCirclePosition = (circleIndex) => {
        //statePosition 0 -> return 50
        //statePosition 1 -> return 150 (margin)
        return (stateSize * (circleIndex + 1)) + calcCircleMargin(circleIndex);
    }

  return (
    <>
        <circle cx={calcCirclePosition(statePosition)} cy={stateSize + topMargin} r={stateSize}></circle>
                <text x={calcCirclePosition(statePosition)} y={stateSize + topMargin}
                    textAnchor="middle"
                    alignmentBaseline="middle"
                    >
                        {state.name}
                </text>

                {isLast ? '' 
                : <line className='arrow' 
                        x1={calcCirclePosition(statePosition) + stateSize} 
                        y1={topMargin + stateSize} 
                        x2={calcCirclePosition(statePosition + 1) - stateSize} 
                        y2={topMargin + stateSize}
                    ></line>
                }
    </>
  )
}
