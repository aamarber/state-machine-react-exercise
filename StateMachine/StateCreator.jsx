import React, { useState } from 'react'

export default function StateCreator({createState}) {
    const [stateName, setStateName] = useState('');

    const clickAddState = () => {
        if(!stateName){
            return;
        }

        createState(stateName);
    };

  return (
    <>
        <input type="text" id='stateName' onChange={(e) => setStateName(e.target.value)} />
        <button id='stateAdder' onClick={clickAddState}>Add State</button>
    </>
  )
}
