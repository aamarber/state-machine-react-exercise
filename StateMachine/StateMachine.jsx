import React, { useState } from 'react'
import StateComponent from './StateComponent';
import StateCreator from './StateCreator';

export default function StateMachine({states}) {

    const [statesInternal, setStates] = useState(states);

    const stateSize = 50;

    const createState = (stateName) => {
        setStates([...statesInternal, {name: stateName}]);
    }

  return (
    <>
        <StateCreator createState={createState}></StateCreator>
        <svg width={(stateSize * statesInternal.length) * 4}>
        {statesInternal.map((state, index) => {
            return <>
                <StateComponent stateSize={50} state={state} statePosition={index} isLast={index === statesInternal.length - 1}></StateComponent>
            </>
        })}
        </svg>
    </>
  )
}
