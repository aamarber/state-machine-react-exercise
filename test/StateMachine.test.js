import renderer from 'react-test-renderer';
import StateMachine from '../StateMachine/StateMachine';

const states = [
    {name: 'A'},
    {name: 'B'},
    {name: 'C'},
    {name: 'D'},
    {name: 'E'},
];

it('creates the state machine', () => {
    const component = renderer.create(
        <StateMachine states={states}></StateMachine>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    
});
