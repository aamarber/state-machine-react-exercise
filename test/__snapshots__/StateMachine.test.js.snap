// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates the state machine 1`] = `
[
  <input
    onChange={[Function]}
    type="text"
  />,
  <button
    onClick={[Function]}
  >
    Add State
  </button>,
  <svg
    width={1000}
  >
    <circle
      cx={50}
      cy={60}
      r={50}
    />
    <text
      alignmentBaseline="middle"
      textAnchor="middle"
      x={50}
      y={60}
    >
      A
    </text>
    <line
      className="arrow"
      x1={100}
      x2={200}
      y1={60}
      y2={60}
    />
    <circle
      cx={250}
      cy={60}
      r={50}
    />
    <text
      alignmentBaseline="middle"
      textAnchor="middle"
      x={250}
      y={60}
    >
      B
    </text>
    <line
      className="arrow"
      x1={300}
      x2={400}
      y1={60}
      y2={60}
    />
    <circle
      cx={450}
      cy={60}
      r={50}
    />
    <text
      alignmentBaseline="middle"
      textAnchor="middle"
      x={450}
      y={60}
    >
      C
    </text>
    <line
      className="arrow"
      x1={500}
      x2={600}
      y1={60}
      y2={60}
    />
    <circle
      cx={650}
      cy={60}
      r={50}
    />
    <text
      alignmentBaseline="middle"
      textAnchor="middle"
      x={650}
      y={60}
    >
      D
    </text>
    <line
      className="arrow"
      x1={700}
      x2={800}
      y1={60}
      y2={60}
    />
    <circle
      cx={850}
      cy={60}
      r={50}
    />
    <text
      alignmentBaseline="middle"
      textAnchor="middle"
      x={850}
      y={60}
    >
      E
    </text>
  </svg>,
]
`;
